geo/src/algorithm/concave_hull.rs

```diff
@@ -0,0 +1,464 @@
+/// let correct_hull = line_string![
+///     (x: 4.0, y: 0.0),
+///     (x: 4.0, y: 4.0),
+///     (x: 0.0, y: 4.0),
+///     (x: 0.0, y: 0.0),
+///     (x: 4.0, y: 0.0),
+/// ];
+///
+/// let res = poly.concave_hull(2.0);
+/// assert_eq!(res.exterior(), &correct_hull);
+/// ```
+pub trait ConcaveHull {
+    type Scalar: CoordinateType;
+    fn concave_hull(&self, concavity: Self::Scalar) -> Polygon<Self::Scalar>;
+}
+
+impl<T> ConcaveHull for Polygon<T>
+where
+    T: Float + RTreeNum + HasKernel,
+{
+    type Scalar = T;
+    fn concave_hull(&self, concavity: Self::Scalar) -> Polygon<Self::Scalar> {
+        let mut points: Vec<_> = self.exterior().0.clone();
+        Polygon::new(concave_hull(&mut points, concavity), vec![])
+    }
+}
+
+impl<T> ConcaveHull for MultiPolygon<T>
+where
+    T: Float + RTreeNum + HasKernel,
+{
+    type Scalar = T;
+    fn concave_hull(&self, concavity: Self::Scalar) -> Polygon<Self::Scalar> {
+        let mut aggregated: Vec<Coordinate<Self::Scalar>> = self
+            .0
+            .iter()
+            .flat_map(|elem| elem.exterior().0.clone())
+            .collect();
+        Polygon::new(concave_hull(&mut aggregated, concavity), vec![])
+    }
+}
+
+impl<T> ConcaveHull for LineString<T>
+where
+    T: Float + RTreeNum + HasKernel,
+{
+    type Scalar = T;
+    fn concave_hull(&self, concavity: Self::Scalar) -> Polygon<Self::Scalar> {
+        Polygon::new(concave_hull(&mut self.0.clone(), concavity), vec![])
+    }
+}
+
+impl<T> ConcaveHull for MultiLineString<T>
+where
+    T: Float + RTreeNum + HasKernel,
+{
+    type Scalar = T;
+    fn concave_hull(&self, concavity: T) -> Polygon<T> {
+        let mut aggregated: Vec<Coordinate<T>> =
+            self.0.iter().flat_map(|elem| elem.0.clone()).collect();
+        Polygon::new(concave_hull(&mut aggregated, concavity), vec![])
+    }
+}
+
+impl<T> ConcaveHull for MultiPoint<T>
+where
+    T: Float + RTreeNum + HasKernel,
+{
+    type Scalar = T;
+    fn concave_hull(&self, concavity: T) -> Polygon<T> {
+        let mut coordinates: Vec<Coordinate<T>> = self.0.iter().map(|point| point.0).collect();
+        Polygon::new(concave_hull(&mut coordinates, concavity), vec![])
+    }
+}
+
+fn find_point_closest_to_line<T>(
+    interior_coords_tree: &RTree<Coordinate<T>>,
+    line: Line<T>,
+    max_dist: T,
+    edge_length: T,
+    concavity: T,
+    line_tree: &RTree<Line<T>>,
+) -> Option<Coordinate<T>>
+where
+    T: Float + RTreeNum,
+{
+    let h = max_dist + max_dist;
+    let w = line.euclidean_length() + h;
+    let two = T::add(T::one(), T::one());
+    let search_dist = T::div(T::sqrt(T::powi(w, 2) + T::powi(h, 2)), two);
+    let centroid = line.centroid();
+    let centroid_coord = Coordinate {
+        x: centroid.x(),
+        y: centroid.y(),
+    };
+    let mut candidates = interior_coords_tree
+        .locate_within_distance(centroid_coord, search_dist)
+        .peekable();
+    let peek = candidates.peek();
```

Comment:
if you add a `?` here, you can avoid doing the big `match` below

```suggestion
    let peek = candidates.peek()?;
```

Reply:
When I try to do this, I get borrow checker errors that I'll need to work through so I'm going to elect to not proceed with this one.